### PR TITLE
Add firebase as fallback for GCM push notifications

### PR DESCRIPTION
**Pull request by pablothedude**
Firebase should be used for fallback notifications if the GCM
notifications fail. This will be used to be able to migrate to the
newer Firebase notification system used for android.

### CODE REVIEW

**Review by joostd — APPROVED** [review]

File: library/tiqr/Tiqr/Message/Firebase.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * This file is part of the tiqr project.
+ * 
+ * The tiqr project aims to provide an open implementation for 
+ * authentication using mobile devices. It was initiated by 
+ * SURFnet and developed by Egeniq.
+ *
+ * More information: http://www.tiqr.org
+ *
+ * @author Joost van Dijk <joost.vandijk@surfnet.nl>
```
Comment (MKodde) [inline library/tiqr/Tiqr/Message/Firebase.php]: Aren't you the author @pablothedude?

File: library/tiqr/Tiqr/Message/Firebase.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * 
+ * @package tiqr
+ *
+ * @license New BSD License - See LICENSE file for details.
+ *
+ * @copyright (C) 2010-2015 SURFnet BV
```
Comment (MKodde) [inline library/tiqr/Tiqr/Message/Firebase.php]: I think the copyright notice should be updated to 2010-2019?

File: library/tiqr/Tiqr/Message/Firebase.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     * Send a message to a device using the firebase API key.
+     *
+     * @param $deviceToken string device ID
+     * @param $alert string alert message
+     * @param $challenge string tiqr challenge url
+     * @param $apiKey string api key for firebase
+     *
+     * @throws Tiqr_Message_Exception_SendFailure
+     */
+    private function _push($deviceToken, $alert, $challenge, $apiKey)
```
Comment (MKodde) [inline library/tiqr/Tiqr/Message/Firebase.php]: This is not your design decision, I'm aware of that. But a message that is also able to send itself is not very solid. 

/observation.

File: library/tiqr/Tiqr/Message/GCM.php [hunk 4]
```diff
@@ -24,9 +24,15 @@
 require_once 'Zend/Mobile/Push/Gcm.php';
 require_once 'Zend/Mobile/Push/Message/Gcm.php';
 
+require_once 'Tiqr/Message/Firebase.php';
+
 /**
  * Android Cloud To Device Messaging message.
  * @author peter
+ *
+ * @WARN This is a locally managed file. This implementation allows for a fallback GCM send event.
```
Comment (MKodde) [inline library/tiqr/Tiqr/Message/GCM.php]: Is this still a locally managed file? Isn't the whole point of this PR that we create a more robust temporary solution for this problem?

**Review by MKodde — COMMENTED** [review]
See some feedback below. Some additional thoughts:

1. Do we need to document the existence `'firebase.apikey'` configuration option somewhere?
2. I'd like to functionally test this change if possible. Please help me in setting things up @pablothedude

File: library/tiqr/Tiqr/Message/Firebase.php (hunk 1 above)
Comment (pablothedude) [reply library/tiqr/Tiqr/Message/Firebase.php]: Removed the line

File: library/tiqr/Tiqr/Message/Firebase.php (hunk 2 above)
Comment (pablothedude) [reply library/tiqr/Tiqr/Message/Firebase.php]: Updated

File: library/tiqr/Tiqr/Message/GCM.php (hunk 4 above)
Comment (pablothedude) [reply library/tiqr/Tiqr/Message/GCM.php]: Remove the comment about the 'locally managed' file

**Review by MKodde — APPROVED** [review]

**Comment by MKodde** [comment]
Thanks for the changes, and I now understand that @joostd will be testing these changes on the test environment. Next step is updating Tiqr I gues?

File: library/tiqr/Tiqr/Message/GCM.php [hunk 5]
```diff
@@ -91,6 +91,7 @@ public function send()
         if ($this->_useFirebaseFallback()) {
             if ($this->_hasErrorResult($response, false)) {
                 $firebase = new Tiqr_Message_Firebase($this->getOptions());
+                $firebase->setCustomProperty('challenge', $this->getCustomProperty('challenge'));
```
Comment (MKodde) [inline library/tiqr/Tiqr/Message/GCM.php]: Good catch!

**Pull request merged**